### PR TITLE
ARROW-6093: [Java] reduce branches in algo for first match in VectorRangeSearcher

### PR DESCRIPTION
**Pull request by liyafan82**
This is a follow up Jira for the improvement suggested by Francois Saint-Jacques in the PR for 
https://github.com/apache/arrow/pull/4925

After careful analysis and proof, I found @fsaintjacques 's proposal really works. The idea behind the algorithm is tricky. This PR is the implementation based on @fsaintjacques 's proposal.

The benefit of this implementation:
1. the code is much clearer
2. there is fewer branches, which is beneficial for performance.

The drawback of this implementation:
1. there can be dumb iterations of the main loop. That is, the while loop may run a few more times after the "solution" is found, just to wait for the loop termination condition is satisfied. 

However, the time complexity of the algorithm is still O(logn), and the number of dumb iterations can not be large.

### CODE REVIEW

File: java/algorithm/src/test/java/org/apache/arrow/algorithm/search/TestVectorRangeSearcher.java [hunk 1]
```diff
@@ -175,4 +183,14 @@ public void testGetUpperBoundsNegative() {
       }
     }
   }
+
+  @Parameterized.Parameters(name = "repeat = {0}")
+  public static Collection<Object[]> getHasher() {
```
Comment (emkornfield) [inline java/algorithm/src/test/java/org/apache/arrow/algorithm/search/TestVectorRangeSearcher.java]: method name seems odd?

**Comment by emkornfield** [comment]
Small nit, otherwise LGTM.

File: java/algorithm/src/test/java/org/apache/arrow/algorithm/search/TestVectorRangeSearcher.java (hunk 1 above)
Comment (liyafan82) [reply java/algorithm/src/test/java/org/apache/arrow/algorithm/search/TestVectorRangeSearcher.java]: nice catch. thank you so much.

**Comment by emkornfield** [comment]
+1, LGTM

**Pull request closed without merging**